File: http/dotnet/httpclient/src/HttpCore.cs
// ------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
// ------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using Microsoft.Kiota.Abstractions;
using Microsoft.Kiota.Abstractions.Serialization;
using Microsoft.Kiota.Abstractions.Store;
using Microsoft.Kiota.Abstractions.Authentication;

namespace Microsoft.Kiota.Http.HttpClient
{
    /// <summary>
    /// The <see cref="IHttpCore"/> implementation for sending requests.
    /// </summary>
    public class HttpCore : IHttpCore, IDisposable
    {
        private readonly System.Net.Http.HttpClient client;
        private readonly IAuthenticationProvider authProvider;
        private IParseNodeFactory pNodeFactory;
        private ISerializationWriterFactory sWriterFactory;
        private readonly bool createdClient;
        /// <summary>
        /// Initializes a new instance of the <see cref="HttpCore"/> class.
        /// <param name="authenticationProvider">The authentication provider.</param>
        /// <param name="parseNodeFactory">The parse node factory.</param>
        /// <param name="serializationWriterFactory">The serialization writer factory.</param>
        /// <param name="httpClient">The native HTTP client.</param>
        /// </summary>
        public HttpCore(IAuthenticationProvider authenticationProvider, IParseNodeFactory parseNodeFactory = null, ISerializationWriterFactory serializationWriterFactory = null, System.Net.Http.HttpClient httpClient = null)
        {
            authProvider = authenticationProvider ?? throw new ArgumentNullException(nameof(authenticationProvider));
            createdClient = httpClient == null;
            client = httpClient ?? HttpClientBuilder.Create(authProvider);
            pNodeFactory = parseNodeFactory ?? ParseNodeFactoryRegistry.DefaultInstance;
            sWriterFactory = serializationWriterFactory ?? SerializationWriterFactoryRegistry.DefaultInstance;
        }
        /// <summary>Factory to use to get a serializer for payload serialization</summary>
        public ISerializationWriterFactory SerializationWriterFactory
        {
            get
            {
                return sWriterFactory;
            }
        }
        /// <summary>
        /// Send a <see cref="RequestInfo"/> instance with a collection instance of <typeparam name="ModelType"></typeparam>
        /// </summary>
        /// <param name="requestInfo">The <see cref="RequestInfo"/> instance to send</param>
        /// <param name="responseHandler">The <see cref="IResponseHandler"/> to use with the response</param>
        public async Task<IEnumerable<ModelType>> SendCollectionAsync<ModelType>(RequestInfo requestInfo, IResponseHandler responseHandler = default) where ModelType : IParsable
        {
            var response = await GetHttpResponseMessage(requestInfo);
            requestInfo.Content?.Dispose();
            if(responseHandler == null)
            {
                var rootNode = await GetRootParseNode(response);
                var result = rootNode.GetCollectionOfObjectValues<ModelType>();
                return result;
            }
            else
                return await responseHandler.HandleResponseAsync<HttpResponseMessage, IEnumerable<ModelType>>(response);
        }
        /// <summary>
        /// Send a <see cref="RequestInfo"/> instance with an instance of <typeparam name="ModelType"></typeparam>
        /// </summary>
        /// <param name="requestInfo">The <see cref="RequestInfo"/> instance to send</param>
        /// <param name="responseHandler">The <see cref="IResponseHandler"/> to use with the response</param>
        public async Task<ModelType> SendAsync<ModelType>(RequestInfo requestInfo, IResponseHandler responseHandler = null) where ModelType : IParsable
        {
            var response = await GetHttpResponseMessage(requestInfo);
            requestInfo.Content?.Dispose();
            if(responseHandler == null)
            {
                var rootNode = await GetRootParseNode(response);
                var result = rootNode.GetObjectValue<ModelType>();
                return result;
            }
            else
                return await responseHandler.HandleResponseAsync<HttpResponseMessage, ModelType>(response);
        }
        /// <summary>
        /// Send a <see cref="RequestInfo"/> instance with a primitive instance of <typeparam name="ModelType"></typeparam>
        /// </summary>
        /// <param name="requestInfo">The <see cref="RequestInfo"/> instance to send</param>
        /// <param name="responseHandler">The <see cref="IResponseHandler"/> to use with the response</param>
        /// <returns></returns>
        public async Task<ModelType> SendPrimitiveAsync<ModelType>(RequestInfo requestInfo, IResponseHandler responseHandler = default)
        {
            var response = await GetHttpResponseMessage(requestInfo);
            requestInfo.Content?.Dispose();
            if(responseHandler == null)
            {
                var modelType = typeof(ModelType);
                if(modelType == typeof(Stream))
                {
                    return (ModelType)(await response.Content.ReadAsStreamAsync() as object);
                }
                else
                {
                    var rootNode = await GetRootParseNode(response);
                    object result;
                    if(modelType == typeof(bool))
                    {
                        result = rootNode.GetBoolValue();
                    }
                    else if(modelType == typeof(string))
                    {
                        result = rootNode.GetStringValue();
                    }
                    else if(modelType == typeof(int))
                    {
                        result = rootNode.GetIntValue();
                    }
                    else if(modelType == typeof(float))
                    {
                        result = rootNode.GetFloatValue();
                    }
                    else if(modelType == typeof(double))
                    {
                        result = rootNode.GetDoubleValue();
                    }
                    else if(modelType == typeof(Guid))
                    {
                        result = rootNode.GetGuidValue();
                    }
                    else if(modelType == typeof(DateTimeOffset))
                    {
                        result = rootNode.GetDateTimeOffsetValue();
                    }
                    else throw new InvalidOperationException("error handling the response, unexpected type");
                    return (ModelType)result;
                }
            }
            else
                return await responseHandler.HandleResponseAsync<HttpResponseMessage, ModelType>(response);
        }
        /// <summary>
        /// Send a <see cref="RequestInfo"/> instance with an empty request body
        /// </summary>
        /// <param name="requestInfo">The <see cref="RequestInfo"/> instance to send</param>
        /// <param name="responseHandler">The <see cref="IResponseHandler"/> to use with the response</param>
        /// <returns></returns>
        public async Task SendNoContentAsync(RequestInfo requestInfo, IResponseHandler responseHandler = null)
        {
            var response = await GetHttpResponseMessage(requestInfo);
            requestInfo.Content?.Dispose();
            if(responseHandler == null)
                response.Dispose();
            else
                await responseHandler.HandleResponseAsync<HttpResponseMessage, object>(response);
        }
        private async Task<IParseNode> GetRootParseNode(HttpResponseMessage response)
        {
            var responseContentType = response.Content.Headers?.ContentType?.MediaType?.ToLowerInvariant();
            if(string.IsNullOrEmpty(responseContentType))
                throw new InvalidOperationException("no response content type header for deserialization");
            using var contentStream = await response.Content.ReadAsStreamAsync();
            var rootNode = pNodeFactory.GetRootParseNode(responseContentType, contentStream);
            response.Dispose();
            return rootNode;
        }
        private async Task<HttpResponseMessage> GetHttpResponseMessage(RequestInfo requestInfo)
        {
            if(requestInfo == null)
                throw new ArgumentNullException(nameof(requestInfo));

            await authProvider.AuthenticateRequestAsync(requestInfo);

            using var message = GetRequestMessageFromRequestInfo(requestInfo);
            var response = await this.client.SendAsync(message);
            if(response == null)
                throw new InvalidOperationException("Could not get a response after calling the service");
            return response;
        }
        private const string ContentTypeHeaderName = "content-type";
        private HttpRequestMessage GetRequestMessageFromRequestInfo(RequestInfo requestInfo)
        {
            var message = new HttpRequestMessage
            {
                Method = new System.Net.Http.HttpMethod(requestInfo.HttpMethod.ToString().ToUpperInvariant()),
                RequestUri = new Uri(requestInfo.URI +
                                        ((requestInfo.QueryParameters?.Any() ?? false) ?
                                            "?" + requestInfo.QueryParameters
                                                        .Select(x => $"{x.Key}{(x.Value == null ? string.Empty : "=")}{x.Value?.ToString() ?? string.Empty}")
                                                        .Aggregate((x, y) => $"{x}&{y}") :
                                            string.Empty)),
            };

            if(requestInfo.MiddlewareOptions.Any())
                requestInfo.MiddlewareOptions.ToList().ForEach(x => message.Options.Set(new HttpRequestOptionsKey<IMiddlewareOption>(x.GetType().FullName), x));
            if(requestInfo.Headers?.Any() ?? false)
                requestInfo.Headers.Where(x => !ContentTypeHeaderName.Equals(x.Key, StringComparison.OrdinalIgnoreCase)).ToList().ForEach(x => message.Headers.Add(x.Key, x.Value));
            if(requestInfo.Content != null)
            {
                message.Content = new StreamContent(requestInfo.Content);
                if(requestInfo?.Headers?.ContainsKey(ContentTypeHeaderName) ?? false)
                    message.Content.Headers.ContentType = new MediaTypeHeaderValue(requestInfo.Headers[ContentTypeHeaderName]);
            }
            return message;
        }
        /// <summary>
        /// Enable the backing store with the provided <see cref="IBackingStoreFactory"/>
        /// </summary>
        /// <param name="backingStoreFactory">The <see cref="IBackingStoreFactory"/> to use</param>
        public void EnableBackingStore(IBackingStoreFactory backingStoreFactory)
        {
            pNodeFactory = ApiClientBuilder.EnableBackingStoreForParseNodeFactory(pNodeFactory) ?? throw new InvalidOperationException("Could not enable backing store for the parse node factory");
            sWriterFactory = ApiClientBuilder.EnableBackingStoreForSerializationWriterFactory(sWriterFactory) ?? throw new InvalidOperationException("Could not enable backing store for the serializer writer factory");
            if(backingStoreFactory != null)
                BackingStoreFactorySingleton.Instance = backingStoreFactory;
        }
        /// <summary>
        /// Dispose/cleanup the client
        /// </summary>
        public void Dispose()
        {
            if(createdClient)
                client?.Dispose();
        }
    }
}

File: http/typescript/fetch/src/httpCore.ts
import { AuthenticationProvider, BackingStoreFactory, BackingStoreFactorySingleton, HttpCore as IHttpCore, Parsable, ParseNodeFactory, RequestInfo, ResponseHandler, ParseNodeFactoryRegistry, enableBackingStoreForParseNodeFactory, SerializationWriterFactoryRegistry, enableBackingStoreForSerializationWriterFactory, SerializationWriterFactory } from '@microsoft/kiota-abstractions';
import { Headers as FetchHeadersCtor } from 'cross-fetch';
import { ReadableStream } from 'web-streams-polyfill';
import { URLSearchParams } from 'url';
import { HttpClient } from './httpClient';
import { Middleware } from "./middleware";
import { FetchOptions } from './fetchOptions';

export interface RequestBuilderConfig {
    parseNodeFactory: ParseNodeFactory,
    serializationWriterFactory: SerializationWriterFactory
}

export interface CoreConfig {
    authenticationProvider: AuthenticationProvider,
    httpClient?: HttpClient,
    middleware?: Middleware | Middleware[],
    fetchOptions?: FetchOptions
}

interface CoreOptions {
    authenticationProvider: AuthenticationProvider,
    httpClient: HttpClient,
    fetchOptions?: FetchOptions
}

export class HttpCore implements IHttpCore {
    public getSerializationWriterFactory(): SerializationWriterFactory {
        return this.requestBuilderConfig.serializationWriterFactory;
    }
    /**
     * @private
     * @constructor
     * Instantiates a new http core service
     */
    private constructor(private requestBuilderConfig: RequestBuilderConfig = { parseNodeFactory: ParseNodeFactoryRegistry.defaultInstance, serializationWriterFactory: SerializationWriterFactoryRegistry.defaultInstance }, private coreOptions: CoreOptions) {
        if (!coreOptions.authenticationProvider) {
            throw new Error('authentication provider cannot be null');
        }
        if (!requestBuilderConfig.parseNodeFactory) {
            throw new Error('parse node factory cannot be null');
        }
        if (!requestBuilderConfig.serializationWriterFactory) {
            throw new Error('serialization writer factory cannot be null');
        }
        if (!coreOptions.httpClient) {
            throw new Error('http client cannot be null');
        }

    }
    private getResponseContentType = (response: Response): string | undefined => {
        const header = response.headers.get("content-type")?.toLowerCase();
        if (!header) return undefined;
        const segments = header.split(';');
        if (segments.length === 0) return undefined;
        else return segments[0];
    }
    public sendCollectionAsync = async <ModelType extends Parsable>(requestInfo: RequestInfo, type: new () => ModelType, responseHandler: ResponseHandler | undefined): Promise<ModelType[]> => {
        if (!requestInfo) {
            throw new Error('requestInfo cannot be null');
        }
        await this.coreOptions.authenticationProvider.authenticateRequest(requestInfo);

        const request = this.getRequestFromRequestInfo(requestInfo);
        const response = await this.coreOptions.httpClient.fetch(this.getRequestUrl(requestInfo), request);
        if (responseHandler) {
            return await responseHandler.handleResponseAsync(response);
        } else {
            const payload = await response.arrayBuffer();
            const responseContentType = this.getResponseContentType(response);
            if (!responseContentType)
                throw new Error("no response content type found for deserialization");

            const rootNode = this.requestBuilderConfig.parseNodeFactory.getRootParseNode(responseContentType, payload);
            const result = rootNode.getCollectionOfObjectValues(type);
            return result as unknown as ModelType[];
        }
    }
    public sendAsync = async <ModelType extends Parsable>(requestInfo: RequestInfo, type: new () => ModelType, responseHandler: ResponseHandler | undefined): Promise<ModelType> => {
        if (!requestInfo) {
            throw new Error('requestInfo cannot be null');
        }
        await this.coreOptions.authenticationProvider.authenticateRequest(requestInfo);

        const request = this.getRequestFromRequestInfo(requestInfo);
        const response = await this.coreOptions.httpClient.fetch(this.getRequestUrl(requestInfo), request);
        if (responseHandler) {
            return await responseHandler.handleResponseAsync(response);
        } else {
            const payload = await response.arrayBuffer();
            const responseContentType = this.getResponseContentType(response);
            if (!responseContentType)
                throw new Error("no response content type found for deserialization");

            const rootNode = this.requestBuilderConfig.parseNodeFactory.getRootParseNode(responseContentType, payload);
            const result = rootNode.getObjectValue(type);
            return result as unknown as ModelType;
        }
    }
    public sendPrimitiveAsync = async <ResponseType>(requestInfo: RequestInfo, responseType: "string" | "number" | "boolean" | "Date" | "ReadableStream", responseHandler: ResponseHandler | undefined): Promise<ResponseType> => {
        if (!requestInfo) {
            throw new Error('requestInfo cannot be null');
        }
        await this.coreOptions.authenticationProvider.authenticateRequest(requestInfo);

        const request = this.getRequestFromRequestInfo(requestInfo);
        const response = await this.coreOptions.httpClient.fetch(this.getRequestUrl(requestInfo), request);
        if (responseHandler) {
            return await responseHandler.handleResponseAsync(response);
        } else {
            switch (responseType) {
                case "ReadableStream":
                    const buffer = await response.arrayBuffer();
                    let bufferPulled = false;
                    const stream = new ReadableStream({
                        pull: (controller) => {
                            if (!bufferPulled) {
                                controller.enqueue(buffer.slice(0))
                                bufferPulled = true;
                            }
                        },
                    });
                    return stream as unknown as ResponseType;
                case 'string':
                case 'number':
                case 'boolean':
                case 'Date':
                    const payload = await response.arrayBuffer();
                    const responseContentType = this.getResponseContentType(response);
                    if (!responseContentType)
                        throw new Error("no response content type found for deserialization");

                    const rootNode = this.requestBuilderConfig.parseNodeFactory.getRootParseNode(responseContentType, payload);
                    if (responseType === 'string') {
                        return rootNode.getStringValue() as unknown as ResponseType;
                    } else if (responseType === 'number') {
                        return rootNode.getNumberValue() as unknown as ResponseType;
                    } else if (responseType === 'boolean') {
                        return rootNode.getBooleanValue() as unknown as ResponseType;
                    } else if (responseType === 'Date') {
                        return rootNode.getDateValue() as unknown as ResponseType;
                    } else {
                        throw new Error("unexpected type to deserialize");
                    }
            }
        }
    }
    public sendNoResponseContentAsync = async (requestInfo: RequestInfo, responseHandler: ResponseHandler | undefined): Promise<void> => {
        if (!requestInfo) {
            throw new Error('requestInfo cannot be null');
        }
        await this.coreOptions.authenticationProvider.authenticateRequest(requestInfo);

        const request = this.getRequestFromRequestInfo(requestInfo);
        const response = await this.coreOptions.httpClient.fetch(this.getRequestUrl(requestInfo), request);
        if (responseHandler) {
            return await responseHandler.handleResponseAsync(response);
        }
    }
    public enableBackingStore = (backingStoreFactory?: BackingStoreFactory | undefined): void => {
        this.requestBuilderConfig.parseNodeFactory = enableBackingStoreForParseNodeFactory(this.requestBuilderConfig.parseNodeFactory);
        this.requestBuilderConfig.serializationWriterFactory = enableBackingStoreForSerializationWriterFactory(this.requestBuilderConfig.serializationWriterFactory);
        if (!this.requestBuilderConfig.serializationWriterFactory || !this.requestBuilderConfig.parseNodeFactory)
            throw new Error("unable to enable backing store");
        if (backingStoreFactory) {
            BackingStoreFactorySingleton.instance = backingStoreFactory;
        }
    }
    private getRequestFromRequestInfo = (requestInfo: RequestInfo): RequestInit => {
        const request = {
            method: requestInfo.httpMethod?.toString(),
            headers: new FetchHeadersCtor(),
            body: requestInfo.content,
        } as RequestInit;
        requestInfo.headers?.forEach((v, k) => (request.headers as Headers).set(k, v));
        return request;
    }
    private getRequestUrl = (requestInfo: RequestInfo): string => {
        let url = requestInfo.URI ?? '';
        if (requestInfo.queryParameters?.size ?? -1 > 0) {
            const queryParametersBuilder = new URLSearchParams();
            requestInfo.queryParameters?.forEach((v, k) => {
                queryParametersBuilder.append(k, `${v}`);
            });
            url = url + '?' + queryParametersBuilder.toString();
        }
        return url;
    }

    /**
     * Factory method to create HttpCore instance for url requests not using RequestBuilders
     * @param coreConfig the configuration used to set authprovider, httpClient or middleware, fetchOptions and more.
     */
    public static createCoreForPrimitiveRequests(coreConfig: CoreConfig): HttpCore {
        const coreOptions: CoreOptions = {
            httpClient: HttpCore.getHttpClient(coreConfig.httpClient, coreConfig.middleware),
            authenticationProvider: coreConfig.authenticationProvider,
            fetchOptions: coreConfig.fetchOptions
        }
        return new HttpCore({ parseNodeFactory: ParseNodeFactoryRegistry.defaultInstance, serializationWriterFactory: SerializationWriterFactoryRegistry.defaultInstance }, coreOptions);
    }

    /**
     * Factory method to create HttpCore instance to support Kiota Request Builders
     * @param requestBuilderConfig the configuration used to set ParseNodeFactory and serializationWriterFactory
     * @param coreConfig the configuration used to set authprovider, httpClient or middleware, fetchOptions and more.
     */
    public static createCoreForRequestBuilders(requestBuilderConfig: RequestBuilderConfig, coreConfig: CoreConfig): HttpCore {
        const coreOptions: CoreOptions = {
            httpClient: HttpCore.getHttpClient(coreConfig.httpClient, coreConfig.middleware),
            authenticationProvider: coreConfig.authenticationProvider,
            fetchOptions: coreConfig.fetchOptions
        }
        return new HttpCore(requestBuilderConfig, coreOptions);
    }

    // TODO : add factory method  return instance with default values.


    /**
     * @private
     * Static method returning an HttpClient instance which will be used by the HttpCore while making request.
     * @param httpClient the an HttpClient instance passed as a configuration of the HttpCore
     * @param middleware a single middleware or an array of middleware used to create an HttpClient
     */
    private static getHttpClient(httpClient?: HttpClient, middleware?: Middleware | Middleware[]): HttpClient {
        if (httpClient && middleware) {
            throw Error("Please provide either the HttpClient instance or the middleware configurations.");
        }
        if (httpClient) {
            return httpClient;
        }
        if (middleware) {
            // TODO: return with new middleware.
            return new HttpClient();
        }
        return new HttpClient();
    }
}